rpcserver.go

```diff
@@ -5583,7 +5580,7 @@ type rpcserverConfig struct {
 }
 
 // newRPCServer returns a new instance of the rpcServer struct.
-func newRPCServer(config *rpcserverConfig) (*rpcServer, error) {
+func newRPCServer(ctx context.Context, config *rpcserverConfig) (*rpcServer, error) {
```

Comment:
Accepting the ctx in both `newRPCServer` and `Run` makes it so that callers need to remember to pass _the same_ context to both.

It's probably better to follow the convention set in other components to only pass the context in `Run()`.

I understand you did that due to the several selects such as in `NotifyBlockConnected` but maybe it would be better to just _not_ perform those sends if the underlying `m.ctx` isn't set (which would be done in `Run()`). The interpretation of that is "if the notification manager isn't running, there's no one to notify".

Basically what I'm thinking of is:

```go
type wsNotificationManager struct {
  //...
  mtx sync.RWMutex
  ctx context.Context
}

func newWsNotificationManager(server *rpcServer) *wsNotificationManager {
//...
}

func (m *wsNotificationManager) run(ctx context.Context) {
  m.mtx.Lock()
  m.ctx = ctx
  m.mtx.Unlock()
  //...
  m.mtx.Lock()
  m.ctx = nil
  m.mtx.Unlock()
}

func (m *wsNotificationManager) NotifyBlockConnected(block *dcrutil.Block) {
  m.mtx.RLock()
  defer m.mtx.RUnlock()
  if m.ctx == nil {
    // Notification manager not started yet. Cannot send notifications.
    return
  }
  //...
}
```

I'm not sure this is worth the effort just to get the API slightly better though...

Reply:
I agree with you, that's better design overall. Will update soon. Thank you.

Reply:
implemented with some modifications, please take another look.